### PR TITLE
DateComponents, URLRequest, URLComponents: Fix hashing

### PR DESCRIPTION
**Pull request by lorentey**
This PR collects fixes previously submitted separately as #1645, #1646 and #1648, fixing the implementation of hashing for the following Foundation types:

- `DateComponents`
- `NSDateComponents`
- `URLRequest`
- `NSURLRequest`
- `URLComponents`
- `NSURLComponents`

Note that this isn't a full audit of Foundation hashing -- this was the bare minimum necessary for #1649. I suspect there may be more types with issues similar to these.

Resolves:
- https://bugs.swift.org/browse/SR-8441
- https://bugs.swift.org/browse/SR-8449
- https://bugs.swift.org/browse/SR-8450

### CODE REVIEW

**Comment by lorentey** [comment]
@swift-ci please test

File: TestFoundation/Utilities.swift [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        Source.self, Target.self,
+        item,
+        { $0.mutableCopy() as! Target },
+        keyPath,
+        values)
+}
+
+// Check that mutating `object` via the specified key path affects its
+// hash value.
+func _checkHashableMutations<Source: Hashable, Target: Hashable, S: Sequence>(
```
Comment (millenomi) [inline TestFoundation/Utilities.swift]: I know these are a ton of arguments and it's essentially fileprivate, but honestly I'd have a more descriptive set of labels because it is _really_ hard to parse the invocations above at a glance.

e.g.: `_checkHashing(of type, matches type, initialValue:, copyHandler:, byMutating keyPaths, throughRange:)`

**Comment by lorentey** [comment]
@swift-ci please test and merge

**Comment by lorentey** [comment]
Oh, swift-ci can’t merge here!

@swift-ci test

**Comment by spevans** [comment]
@lorentey There is an Xcode project, `DarwinCompatibilityTests` in the toplevel that can be used to run the `TestFoundation` tests against Darwin's native foundation. The tests should show up in the unit test pane, although not all of them currently pass.

**Comment by lorentey** [comment]
I managed to run these tests on Darwin's Foundation, and filed rdar://problem/43422017 with my findings. While Darwin Foundation (erroneously) neglects to mix some of the properties that affect equality into the hash value, I believe there is no need for swift-corelibs-foundation to emulate this behavior. (Emulating exact hash values would not be a worthy exercise anyway, and given the lack of random seeding, it may even be actively harmful to do so in server contexts.) So I think the definitions can stay as defined here.

**Comment by lorentey** [comment]
Now if only I could get PR testing to update GitHub status... The tests seem to finish successfully, but the PR stays yellow.

@swift-ci test
